jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java

```diff
@@ -95,166 +93,189 @@ public MetaInfConfiguration()
-                    {
-                        for (Path listEntry: listing.toList())
-                        {
-                            Resource resource = resourceFactory.newResource(listEntry);
-                            context.getMetaData().addContainerResource(resource);
-                        }
-                    }
-                }
-                else
-                {
-                    Resource resource = resourceFactory.newResource(path);
-                    context.getMetaData().addContainerResource(resource);
-                }
-            }
+            cache.putIfAbsent(target, fragment);
+            return new FragmentMapping(target, fragment);
         }
 
-        if (LOG.isDebugEnabled())
-            LOG.debug("Container paths selected:{}", context.getMetaData().getContainerResources());
+        return null;
     }
 
-    /**
-     * Finds the jars that are either physically or virtually in
-     * WEB-INF/lib, and applies an optional filter to their full
-     * pathnames.
-     * <p>
-     * The filter selects which jars will later be examined for META-INF
-     * information and annotations. If there is no pattern, then
-     * all jars are considered selected.
-     *
-     * @param context the WebAppContext being deployed
-     */
-    public void findAndFilterWebAppPaths(WebAppContext context)
-        throws Exception
+    private List<Resource> scanMetaInfResources(Stream<Resource> targets, Map<Resource, Resource> cache)
     {
-        //Apply filter to WEB-INF/lib jars
-        String pattern = (String)context.getAttribute(WEBINF_JAR_PATTERN);
-        ResourceUriPatternPredicate webinfPredicate = new ResourceUriPatternPredicate(pattern, true);
+        assert cache != null;
 
-        List<Resource> jars = findJars(context);
-        if (LOG.isDebugEnabled())
-            LOG.debug("webapp {}={} jars {}", WEBINF_JAR_PATTERN, pattern, jars);
+        return targets
+            .map(target -> getMetaInfResourceDir(target, cache))
+            .filter(Objects::nonNull)
+            .toList();
+    }
 
-        // Only add matching Resources to metadata.webInfResources
-        if (jars != null)
+    private Resource getMetaInfResourceDir(Resource target, Map<Resource, Resource> cache)
+    {
+        Resource dir = cache.get(target);
+        if (Resources.isReadableDirectory(dir))
+            return dir;
+
+        dir = target.resolve("META-INF/resources");
+        if (Resources.isReadableDirectory(dir))
         {
-            jars.stream()
-                .filter(webinfPredicate)
-                .forEach(resource -> context.getMetaData().addWebInfResource(resource));
+            cache.putIfAbsent(target, dir);
+            return dir;
         }
+        return null;
     }
 
-    protected List<URI> getAllContainerJars(final WebAppContext context)
+    /**
+     * Only look for Servlet 3+ features ({@code META-INF/web-fragment.xml} and {@code META-INF/resources})
```

Comment:
This is not the right algorithm: we only need to skip fragments iff metadata-complete or < servlet3 or not discovered configuration. The current code is correct:
```java
        if (context.getMetaData().isMetaDataComplete() || (context.getServletContext().getEffectiveMajorVersion() < 3) && !context.isConfigurationDiscovered())
            scanTypes.remove(METAINF_FRAGMENTS);
```